apps/src/lib/levelbuilder/code-docs-editor/ProgrammingMethodEditor.jsx

```diff
@@ -0,0 +1,207 @@
+import PropTypes from 'prop-types';
+import React, {useState} from 'react';
+import OrderableList from './OrderableList';
+import ExampleEditor from './ExampleEditor';
+import ParameterEditor from './ParameterEditor';
+import TextareaWithMarkdownPreview from '@cdo/apps/lib/levelbuilder/TextareaWithMarkdownPreview';
+import CollapsibleEditorSection from '@cdo/apps/lib/levelbuilder/CollapsibleEditorSection';
+import HelpTip from '@cdo/apps/lib/ui/HelpTip';
+import SaveBar from '@cdo/apps/lib/levelbuilder/SaveBar';
+import {createUuid, navigateToHref} from '@cdo/apps/utils';
+import $ from 'jquery';
+import color from '@cdo/apps/util/color';
+
+function useProgrammingMethod(initialProgrammingMethod) {
+  const [programmingMethod, setProgrammingMethod] = useState(
+    initialProgrammingMethod
+  );
+
+  function updateProgrammingMethod(key, value) {
+    setProgrammingMethod({...programmingMethod, [key]: value});
+  }
+
+  return [programmingMethod, updateProgrammingMethod, setProgrammingMethod];
+}
+
+function renderExampleEditor(example, updateFunc) {
+  return <ExampleEditor example={example} updateExample={updateFunc} />;
+}
+
+function renderParameterEditor(parameter, updateFunc) {
+  return <ParameterEditor parameter={parameter} update={updateFunc} />;
+}
+
+export default function ProgrammingMethodEditor({initialProgrammingMethod}) {
+  // We don't want to update id or key
+  const {id, key, ...remainingProgrammingMethod} = initialProgrammingMethod;
+  if (remainingProgrammingMethod.examples) {
+    remainingProgrammingMethod.examples.forEach(e => (e.key = createUuid()));
+  }
+  if (remainingProgrammingMethod.parameters) {
+    remainingProgrammingMethod.parameters.forEach(p => (p.key = createUuid()));
+  }
```

Comment:
[Here](https://github.com/code-dot-org/code-dot-org/blob/b2e1c5f1cd3da6dd57ca87dbebccec7573f3306a/apps/src/lib/levelbuilder/code-docs-editor/OrderableList.jsx#L38). But implicitly, I was talking about Rails' requirement that when a list of items are rendered that each must have a unique `key`.

Reply:
(Did you mean "React's requirement..."?)

To make sure I understand the intent here, is the following accurate: The passed in `initialProgrammingMethod` is used to initialize the state variable (and then is not used for subsequent renders).  Before we initial the state variable, we want to add a unique id to each element because that is required by OrderableList.

If that's the case, why do we extract `id` and `key` here?  Should we extract it from `programmingMethod` below instead?

Also, a potential simplification would be to just assign the array index as the id (up to you).

Reply:
🤦‍♀️ yes I did mean React.

> The passed in initialProgrammingMethod is used to initialize the state variable (and then is not used for subsequent renders). Before we initial the state variable, we want to add a unique id to each element because that is required by OrderableList.

Yes

> If that's the case, why do we extract id and key here? Should we extract it from programmingMethod below instead?

I don't understand what you mean here? They're extracted because they shouldn't change here but if that part is confusing, I can add them to the state.

> Also, a potential simplification would be to just assign the array index as the id (up to you).

I could do that! And that probably is clearer. Mind if I do that in a follow-up so I can update a bunch of places at once?

Reply:
Discussed offline and came to two action items for me:
1. Update the comment on line 39 to explain why these properties are not stored in state
2. Move these setup lines to a function that is passed into `useState` ([React documentation](https://reactjs.org/docs/hooks-reference.html#lazy-initial-state)).

Both of these issues exist across multiple components so I'm going to do both as a follow up.